connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/inbound/DefaultProcessInstanceContext.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
+ * under one or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information regarding copyright
+ * ownership. Camunda licenses this file to you under the Apache License,
+ * Version 2.0; you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.camunda.connector.runtime.core.inbound;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.camunda.connector.api.inbound.ProcessInstanceContext;
+import io.camunda.connector.api.validation.ValidationProvider;
+import io.camunda.connector.api.validation.ValidationUtil;
+import io.camunda.connector.feel.FeelEngineWrapper;
+import io.camunda.connector.feel.FeelEngineWrapperException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Supplier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class DefaultProcessInstanceContext implements ProcessInstanceContext {
+  private static final Logger LOGGER = LoggerFactory.getLogger(DefaultProcessInstanceContext.class);
+
+  private final Long processInstanceKey;
+  private final ValidationProvider validationProvider;
+  private final FeelEngineWrapper feelEngineWrapper;
+  private final ObjectMapper objectMapper;
+  private final Map<String, Object> contextProperties;
+  private final Supplier<Map<String, Object>> operatePropertiesSupplier;
+
+  public DefaultProcessInstanceContext(
+      final Long processInstanceKey,
+      final ValidationProvider validationProvider,
+      final FeelEngineWrapper feelEngineWrapper,
+      final ObjectMapper objectMapper,
+      final Map<String, Object> contextProperties,
+      final Supplier<Map<String, Object>> operateVariables) {
+    this.processInstanceKey = processInstanceKey;
+    this.validationProvider =
+        validationProvider == null
+            ? ValidationUtil.discoverDefaultValidationProviderImplementation()
+            : validationProvider;
+    this.feelEngineWrapper = feelEngineWrapper;
+    this.objectMapper = objectMapper;
+    this.contextProperties = contextProperties;
+    this.operatePropertiesSupplier = operateVariables;
+  }
+
+  @Override
+  public Long getKey() {
+    return processInstanceKey;
+  }
+
+  @Override
+  public <T> T bind(final Class<T> cls) {
```

Comment:
As discussed, please replace the `bind` method with `getVariables()` and add a simple check whether the variables have been fetched already.

Reply:
Not really as its about creating a `bind` method that works using jackson. We wanted to use `getVariables()` now to avoid this custom FEEL handling code we have here.

Reply:
@Oleksiivanov You dont as you dont need validation for process instance variable binding.

Reply:
You can create a record class that has a Function<Map<String,Object>> and bind it to the url. Then call it with the `getVariables()` result to use it. You dont have to use the Http Request class 1:1. Just pass/map the values.

Reply:
As discussed yesterday we keep the bind method and refactor later.